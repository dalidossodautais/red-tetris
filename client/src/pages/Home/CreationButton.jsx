import React, { useEffect, useState } from "react";
import styled from "styled-components";
import Button from "../../components/inputs/Button";
import Select from "../../components/inputs/Select";
import TextField from "../../components/inputs/TextField";
import Title from "../../components/texts/Title";
import Modal from "../../components/Modal";
import Toolbar from "../../components/Toolbar";
import { createRoom, handleCreation, unhandleCreation } from "../../middleware";

const InputContainer = styled.div`
  display: flex;
  height: 50px;
  margin-left: 30px;
  margin-top: 15px;
`;

const FooterContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export default function CreationButton() {
  const [error, setError] = useState(null);
  const [name, setName] = useState("");
  const [password, setPassword] = useState("");
  const [type, setType] = useState("solo");
  const [isOpened, setOpened] = useState(false);
  useEffect(() => {
    handleCreation(setError);
    return () => {
      unhandleCreation();
    };
  }, [setError]);
  useEffect(() => {
    return () => {
      setError(null);
      setName("");
    };
  }, [isOpened, setError, setName]);
  return (
    <>
      <Button
        onClick={() => {
          setOpened(true);
        }}
        variant="outlined"
      >
        Create
      </Button>
      <Modal
        onClose={() => setOpened(false)}
        onKeyUp={(event) => {
          if (event.key === "Enter")
            createRoom({ name, password, type }, setError);
        }}
        open={isOpened}
      >
        <Toolbar marginbottom="10px">
          <Title fontWeight="normal">Choose a room name</Title>
        </Toolbar>
        <InputContainer>
          <Select
            choices={[
              { label: "Solo", value: "solo" },
              { label: "Public", value: "public" },
              { label: "Private", value: "private" },
            ]}
            label="Gamemode"
            onChange={(event) => {
              setType(event.target.value);
            }}
            value={type}
            variant="outlined"
          />
          <TextField
            error={error && error.name ? error.name : null}
            hide={!["public", "private"].includes(type)}
            label="Room name"
            onChange={(event) => {
              setName(
                event.target.value.replace(/[^0-9A-Za-z _-]/g, "").slice(0, 20)
              );
              setError(null);
            }}
            value={name}
            variant="outlined"
          />
          <TextField
            error={error && error.password ? error.password : null}
            hide={type !== "private"}
            label="Password"
            onChange={(event) => {
              setPassword(event.target.value.slice(0, 20));
              setError(null);
            }}
            type="password"
            value={password}
            variant="outlined"
          />
        </InputContainer>
        <FooterContainer>
          <Button
            disabled={
              !!error ||
              (["public", "private"].includes(type) && name.length < 3) ||
              (type === "private" && password.length < 3)
            }
            onClick={() => createRoom({ name, password, type }, setError)}
            variant="contained"
          >
            Create
          </Button>
        </FooterContainer>
      </Modal>
    </>
  );
}
