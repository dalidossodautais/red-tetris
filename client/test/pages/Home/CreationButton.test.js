import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import CreationButton from "../../../src/pages/Home/CreationButton";

describe("CreationButton", () => {
  test("Standard", () => {
    const renderer = new ShallowRenderer();
    renderer.render(<CreationButton />);
    const result = renderer.getRenderOutput();
    // rendering
    expect(result).toMatchSnapshot();
    // re-rendering
    expect(result).toMatchSnapshot();
    expect(result.props.children[0]).toMatchSnapshot();
  });
});
