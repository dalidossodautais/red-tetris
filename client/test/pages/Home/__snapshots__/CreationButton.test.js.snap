// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreationButton Standard 1`] = `
<React.Fragment>
  <_default
    onClick={[Function]}
    variant="outlined"
  >
    Create
  </_default>
  <_default
    onClose={[Function]}
    onKeyUp={[Function]}
    open={false}
  >
    <Styled(WithStyles(ForwardRef(Toolbar)))
      marginbottom="10px"
    >
      <styled.h2
        fontWeight="normal"
      >
        Choose a room name
      </styled.h2>
    </Styled(WithStyles(ForwardRef(Toolbar)))>
    <styled.div>
      <_default
        choices={
          Array [
            Object {
              "label": "Solo",
              "value": "solo",
            },
            Object {
              "label": "Public",
              "value": "public",
            },
            Object {
              "label": "Private",
              "value": "private",
            },
          ]
        }
        label="Gamemode"
        onChange={[Function]}
        value="solo"
        variant="outlined"
      />
      <_default
        error={null}
        hide={true}
        label="Room name"
        onChange={[Function]}
        value=""
        variant="outlined"
      />
      <_default
        error={null}
        hide={true}
        label="Password"
        onChange={[Function]}
        type="password"
        value=""
        variant="outlined"
      />
    </styled.div>
    <styled.div>
      <_default
        disabled={false}
        onClick={[Function]}
        variant="contained"
      >
        Create
      </_default>
    </styled.div>
  </_default>
</React.Fragment>
`;

exports[`CreationButton Standard 2`] = `
<React.Fragment>
  <_default
    onClick={[Function]}
    variant="outlined"
  >
    Create
  </_default>
  <_default
    onClose={[Function]}
    onKeyUp={[Function]}
    open={false}
  >
    <Styled(WithStyles(ForwardRef(Toolbar)))
      marginbottom="10px"
    >
      <styled.h2
        fontWeight="normal"
      >
        Choose a room name
      </styled.h2>
    </Styled(WithStyles(ForwardRef(Toolbar)))>
    <styled.div>
      <_default
        choices={
          Array [
            Object {
              "label": "Solo",
              "value": "solo",
            },
            Object {
              "label": "Public",
              "value": "public",
            },
            Object {
              "label": "Private",
              "value": "private",
            },
          ]
        }
        label="Gamemode"
        onChange={[Function]}
        value="solo"
        variant="outlined"
      />
      <_default
        error={null}
        hide={true}
        label="Room name"
        onChange={[Function]}
        value=""
        variant="outlined"
      />
      <_default
        error={null}
        hide={true}
        label="Password"
        onChange={[Function]}
        type="password"
        value=""
        variant="outlined"
      />
    </styled.div>
    <styled.div>
      <_default
        disabled={false}
        onClick={[Function]}
        variant="contained"
      >
        Create
      </_default>
    </styled.div>
  </_default>
</React.Fragment>
`;

exports[`CreationButton Standard 3`] = `
<_default
  onClick={[Function]}
  variant="outlined"
>
  Create
</_default>
`;
